src/hackney_happy.erl

```diff
@@ -0,0 +1,77 @@
+-module(hackney_happy).
+
+-export([connect/3, connect/4]).
+
+-include("hackney_internal.hrl").
+-include_lib("kernel/include/inet.hrl").
+
+-define(TIMEOUT, 250).
+-define(CONNECT_TIMEOUT, 5000).
+
+connect(Hostname, Port, Opts) ->
+  connect(Hostname, Port, Opts, ?CONNECT_TIMEOUT).
+
+connect(Hostname, Port, Opts, Timeout) ->
+  case hackney_cidr:is_ipv6(Hostname) of
+    true ->
+      ?report_debug("connect using IPv6", [{hostname, Hostname}, {port, Port}]),
+      gen_tcp:connect(Hostname, Port, [inet6 | Opts], Timeout);
+    false ->
+      case hackney_cidr:is_ipv4(Hostname) of
+        true ->
+          ?report_debug("connect using IPv4", [{hostname, Hostname}, {port, Port}]),
+          gen_tcp:connect(Hostname, Port, [inet | Opts], Timeout);
+        false ->
+          ?report_debug("happy eyeballs, try to connect using IPv6",  [{hostname, Hostname}, {port, Port}]),
+          Self = self(),
+          Addrs = getaddrs(Hostname),
+          Pid = spawn_link( fun() -> try_connect(Addrs, Port, Opts, Self, {error, nxdomain}) end),
+          MRef = erlang:monitor(process, Pid),
+          receive
+            {happy_connect, OK} ->
+              erlang:demonitor(MRef, [flush]),
+              OK;
+            {'DOWN', MRef, _Type, _Pid, Info} ->
+              {'error', {'connect_error', Info}}
+          after Timeout -> 
+                  erlang:demonitor(MRef, [flush]),
+                  {error, connect_timeout}
+          end
+      end
+  end.
+ 
+getaddrs(Hostname) ->
+  IP6Addrs = [{Addr, 'inet6'} || Addr <- getbyname(Hostname, 'aaaa')],
+  IP4Addrs = [{Addr, 'inet'} || Addr <- getbyname(Hostname, 'a')],
+  IP6Addrs ++ IP4Addrs.
```

Comment:
👋 @benoitc 

I wonder if it would make sense to "intersperse" the addresses here like [ipv6, ipv4, ipv6, ipv4, etc.]?

Reply:
Sort of like this: https://github.com/ruslandoga/happy_tcp/blob/afbf889d566e6de743f66978f44b831987acbd6e/lib/happy_tcp.ex#L61-L81